### PR TITLE
Bug #565 xASL_adm_DeleteFileList symlink

### LINKED ISSUE #565: Behavior xASL_adm_DeleteFileList for symbolic links
### Release Notes
* Improved behavior of `xASL_adm_DeleteFileList`, which now understands if the same file is tried to delete twice (e.g., in the case of symbolic links)

### Problem statement
Symbolic links are found by `xASL_adm_GetFileList`, if they point to a folder they are found regardless of the `bGetDirNames` input argument. This can become problematic if the retrieved list includes both the original files and the symbolic links pointing to these files. An example of this would be in xASL_adm_DeleteFileList` at line 62 `delete(filepaths{:});` where a full list is deleted at once for improved performance, but if this list includes both the original file and the symbolic link pointing to the original file, this may induce conflicts.

An elegant solution could be that `xASL_adm_GetFileList` understands if a file that it found (either directly or as symbolic link within a symbolic folder link) point to a file that it already found in the list. This would avoid processing the same file multiple times.

This behavior was detected by trying BASIL's implementation in `xASL_quant_Basil`, which performs an iteration, creates folders with the names step_1 step_2 ... step_n and then always has the same symbolic link folder pointing to the latest iteration, called step_n_latest.

### PR DESCRIPTION
**Pull request by HenkMutsaerts**
### Linked issue

Closes #565

### Notes for reviewing
Please check the issue, thanks!

### CODE REVIEW

**Review by jan-petr — COMMENTED** [review]
Why do we use the delete function instead of calling xASL_delete one by one in the first place? Is that significantly faster?

**Comment by HenkMutsaerts** [comment]
> Why do we use the delete function instead of calling xASL_delete one by one in the first place? Is that significantly faster?

If you call `delete`, it will delete in one batch, which crashes if you have symbolic links inside the list of which the directory above it was also in the list but is already deleted. So `delete` crashes if you have a list with files that are also linked within a symbolic link folder. This is the case for BASILs output. BASIL outputs each fit iteration in a separate subfolder, and creates a symbolic link to the latest folder, assuming this is the most optimal output.

**Comment by MichaelStritt** [comment]
@HenkMutsaerts: Should I just check if I can delete a list of files and if that works it's okay? Or does this require more extensive testing?

**Review by MichaelStritt — CHANGES_REQUESTED** [review]
Removing files seems to work and is fast, but the file list that is returned is empty?

```matlab
>> pathTest = '.\testFolderToRemove';
>> testFiles = xASL_adm_GetFsList(pathTest, '^.+\.exe$');
>> testFiles

testFiles =

  1×8 cell array

    {'myFile (1).exe'}    {'myFile (2).exe'}    {'myFile (3).exe'}    {'myFile (4).exe'}    
    {'myFile (5).exe'}    {'myFile (6).exe'}    {'myFile (7).exe'}    {'myFile (8).exe'}

>> removedFiles = xASL_adm_DeleteFileList(pathTest, '^.+\.exe$');
>> removedFiles

removedFiles =

  0×0 empty char array
```

**Comment by jan-petr** [comment]
> > Why do we use the delete function instead of calling xASL_delete one by one in the first place? Is that significantly faster?
> 
> If you call `delete`, it will delete in one batch, which crashes if you have symbolic links inside the list of which the directory above it was also in the list but is already deleted. So `delete` crashes if you have a list with files that are also linked within a symbolic link folder. This is the case for BASILs output. BASIL outputs each fit iteration in a separate subfolder, and creates a symbolic link to the latest folder, assuming this is the most optimal output.

Yes. I understand why `delete` crashes. But I don't understand why we have the try-catch statement with catch calling xASL_delete one by one - why don't we skip the use of `delete` and delete only using the `xASL_delete` in the catch branch - would that be so slower?

I am simply looking for the motivation to try lines 66-67 and not going right away for solution on line 68-79? Is that faster?

**Comment by HenkMutsaerts** [comment]
Yes, delete is faster, that was the only motivation for the try

**Comment by jan-petr** [comment]
> Yes, delete is faster, that was the only motivation for the try

OK. That's a satisfactory answer :)

**Review by jan-petr — APPROVED** [review]
OK.

**Comment by MichaelStritt** [comment]
## Develop

### Test without space in path

```
>> pathTest = '.\Server_xASL\sdfsdf';
>> filepaths = xASL_adm_DeleteFileList(pathTest, '^.+$');
>> filepaths
filepaths =
  4×1 cell array
    {'.\Server_xASL\sdfsdf\Git-2.32.0-64-bit - Kopie (2).exe'    }
    {'.\Server_xASL\sdfsdf\Git-2.32.0-64-bit - Kopie - Kopie.exe'}
    {'.\Server_xASL\sdfsdf\Git-2.32.0-64-bit - Kopie.exe'        }
    {'.\Server_xASL\sdfsdf\Git-2.32.0-64-bit.exe'                }
```

### Test with space in path

```
>> pathTest = '.\Server_xASL\sdf sdf';
>> filepaths = xASL_adm_DeleteFileList(pathTest, '^.+$');
>> filepaths
filepaths =
  4×1 cell array
    {'.\sdf sdf\Git-2.32.0-64-bit - Kopie (2).exe'    }
    {'.\Server_xASL\sdf sdf\Git-2.32.0-64-bit - Kopie - Kopie.exe'}
    {'.\Server_xASL\sdf sdf\Git-2.32.0-64-bit - Kopie.exe'        }
    {'.\Server_xASL\sdf sdf\Git-2.32.0-64-bit.exe'                }
```

## Feature branch

### Test without space in path

```
>> pathTest = '.\Server_xASL\sdfsdf';
>> filepaths = xASL_adm_DeleteFileList(pathTest, '^.+$');
>> filepaths
filepaths =
  0×0 empty char array
```

### Test with space in path

```
>> pathTest = '.\Server_xASL\sdf sdf';
>> filepaths = xASL_adm_DeleteFileList(pathTest, '^.+$');
>> filepaths
filepaths =
  0×0 empty char array
```

## Conclusion

@HenkMutsaerts: Seems like this bug is not in develop and is not related to spaces in file names.

**Comment by HenkMutsaerts** [comment]
@MichaelStritt can you retest? I indeed reinitialized the output value `filepaths` through the new code, which would then find any residual files to be deleted but this would also render this list empty if the requested files were correctly deleted. Now this behavior is corrected, and another output parameters `pathsNotDeleted` list added for clarity. Is it now working correctly at your end?

**Review by MichaelStritt — APPROVED** [review]
Seems to work now 👍 

```
>> pathWithoutSpaces = '.\Server_xASL\testtest';
>> [deletedFiles,NotDeletedFiles] = xASL_adm_DeleteFileList(pathWithoutSpaces, '^.+$')

deletedFiles =

  4×1 cell array

    {'.\Server_xASL\testtest\Git-2.32.0-64-bit - Kopie (2).exe'    }
    {'.\Server_xASL\testtest\Git-2.32.0-64-bit - Kopie - Kopie.exe'}
    {'.\Server_xASL\testtest\Git-2.32.0-64-bit - Kopie.exe'        }
    {'.\Server_xASL\testtest\Git-2.32.0-64-bit.exe'                }

NotDeletedFiles =
  0×0 empty char array
```

```
>> pathWithSpaces = '.\Server_xASL\test test';
>> [deletedFiles,NotDeletedFiles] = xASL_adm_DeleteFileList(pathWithSpaces, '^.+$')

deletedFiles =
  4×1 cell array

    {'.\Server_xASL\test test\Git-2.32.0-64-bit - Kopie (2).exe'    }
    {'.\Server_xASL\test test\Git-2.32.0-64-bit - Kopie - Kopie.exe'}
    {'.\Server_xASL\test test\Git-2.32.0-64-bit - Kopie.exe'        }
    {'.\Server_xASL\test test\Git-2.32.0-64-bit.exe'                }


NotDeletedFiles =
  0×0 empty char array
```

**Review by MichaelStritt — CHANGES_REQUESTED** [review]
So the deleted files list seems to work, the other one seems to be bugged though. Check out my UnitTest. You should be able to run it pretty easily.

File: Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+fclose('all');
+
+
+% Run your test here
+pathWithSpaces = fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces');
+pathWithoutSpaces = fullfile(TestRepository,'UnitTesting','working_directory','dir_without_spaces');
+[deletedFilesA,NotDeletedFilesA] = xASL_adm_DeleteFileList(pathWithSpaces, '^.+.txt$');
+[deletedFilesB,NotDeletedFilesB] = xASL_adm_DeleteFileList(pathWithoutSpaces, '^.+.txt$');
+
+rmdir(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces'), 's');
```
[the remark this answers is not in the document]
Comment (MichaelStritt) [reply Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m:57]: shouldn't be necessary, the basic folder structure should be there if the testing repository is there

File: Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+% Give your individual subtest a name
+UnitTest.tests(1).testname = 'Delete file lists (basic test)';
+
+% Start the test
+testTime = tic;
+
+% Prepare tests
+
+% Create six test files in test directory
+mkdir(fullfile(TestRepository,'UnitTesting','working_directory'),'dir with spaces');
```
[the remark this answers is not in the document]
Comment (MichaelStritt) [reply Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m:31]: Good idea

File: Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+fid01 = fopen(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces','test_1.txt'), 'wt');
+fid02 = fopen(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces','test_2.txt'), 'wt');
+fid03 = fopen(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces','test_3.txt'), 'wt');
+fid04 = fopen(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces','dont_1.json'), 'wt');
+fid05 = fopen(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces','dont_2.json'), 'wt');
+fid06 = fopen(fullfile(TestRepository,'UnitTesting','working_directory','dir with spaces','dont_3.json'), 'wt');
+fclose('all');
+
+% Create six test files in test directory
```
[the remark this answers is not in the document]
Comment (MichaelStritt) [reply Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m:40]: No, it should be six files? Three txt files and three json files.

File: Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION:  Should be run using xASL_ut_UnitTesting.
+%
+% EXAMPLE:      UnitTests(1) = xASL_ut_function_xASL_adm_DeleteFileList(TestRepository);
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% Copyright 2015-2021 ExploreASL
+
+
+%% Test run 1
```
[the remark this answers is not in the document]
Comment (MichaelStritt) [reply Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m:20]: You can define multiple tests so there are multiple runs of one function which is being tested.

File: Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelStritt) [reply Testing/UnitTests/xASL_ut_function_xASL_adm_DeleteFileList.m]: b87f04a

**Review by MichaelStritt — APPROVED** [review]
Sorry, I should have read the header more carefully. I thought `pathsNotDeleted` would be the paths in the directory that weren't deleted. You describe it correctly though 👍

**Pull request merged**